tests/PHPStan/Analyser/data/bug-4875.php

```diff
@@ -30,8 +30,8 @@ function doFoo()
 		$mock = $this->mockIt(Blah::class);
 
 		assertType('Bug4875\Blah&Bug4875\Mock', $mock);
-		assertType('class-string<Bug4875\Blah>&class-string<Bug4875\Mock>', $mock::class);
-		assertType('class-string<Bug4875\Blah>&class-string<Bug4875\Mock>', get_class($mock));
+		assertType('class-string<Bug4875\Blah&Bug4875\Mock>', $mock::class);
+		assertType('class-string<Bug4875\Blah&Bug4875\Mock>', get_class($mock));
```

Comment:
Can you please explain this a bit? Why `class-string<Bug4875\Blah>&class-string<Bug4875\Mock>` and `class-string<Bug4875\Blah&Bug4875\Mock>` aren't equivalent and why this change fixes https://github.com/phpstan/phpstan/issues/7987? Thanks!

Reply:
The most important fix is collapsing `class-string<T of A|B>&class-string<A>` into `class-string<T of A>`. This allows inferring `T` more reliably.

Semantically they represent the same thing, and since the collapsed form improves template type inference I think it's better to collapse it.

Reply:
thank you both! I will test soon and report back if there are any issues